File: piper/android/app/src/main/java/io/gopiper/piper/engine/statements/utils_tryCatch.kt
package io.gopiper.piper.engine.statements

import android.util.Log
import io.gopiper.piper.engine.Piper
import org.json.JSONObject
import java.lang.Exception

suspend fun Piper.executeUtilsTryCatch(statement: JSONObject, onError: (e: Exception) -> Unit) {
    val tryBlock: String
    val catchBlock: String

    try {
        val pointer = statement.getJSONObject("pointer")
        tryBlock = pointer["try"].toString()
        catchBlock = pointer["catch"].toString()
    } catch (e: Exception) {
        onError(e)
        return
    }

    try {
        canShowErrorAlert = false
        execute(tryBlock)
        canShowErrorAlert = true
        if (isTerminated) {
            isTerminated = false
            throw Exception()
        }
    } catch (e: Exception) {
        try {
            execute(catchBlock)
        } catch (e: Exception) {
            onError(e)
        }
    }
}

File: piper/extensions/chrome/src/engine/statements/utils_tryCatch.js
import { stringify } from "../util"


export default async function(statement, onError) {
    let tryBlock
    let catchBlock
    try {
        let pointer = statement.pointer
        tryBlock = stringify(pointer.try)
        catchBlock = stringify(pointer.catch)
    } catch (e) {
        onError(e)
        return
    }


    try {
        this.canShowErrorAlert = false
        await this.execute(tryBlock)
        this.canShowErrorAlert = true
        if (this.isTerminated) {
            this.isTerminated = false
            throw new Error("Try failed")
        }
    } catch(e) {
        try {
            await this.execute(catchBlock)
        } catch (e) {
            onError(e)
        }
    }
}
